File: 2023-2024/sourcecode/mod3/lab9/RepositoryBasicsJava/app/src/main/java/com/example/repositorybasicsjava/ui/UserFragment.java
package com.example.repositorybasicsjava.ui;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.example.repositorybasicsjava.R;
import com.example.repositorybasicsjava.databinding.FragmentUserBinding;
import com.example.repositorybasicsjava.viewmodel.UserViewModel;

import java.util.ArrayList;

public class UserFragment extends Fragment {

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        com.example.repositorybasicsjava.databinding.FragmentUserBinding binding = FragmentUserBinding.inflate(inflater);

        UserViewModel viewModel = new ViewModelProvider(this).get(UserViewModel.class);
        UserAdapter userAdapter = new UserAdapter(new UserComparator());

        viewModel.getUsersList().observe(getViewLifecycleOwner(), users -> userAdapter.submitList(new ArrayList<>(users)));

        binding.rvList.setAdapter(userAdapter);
        binding.rvList.setLayoutManager(new LinearLayoutManager(requireContext()));

        return binding.getRoot();
    }
}

File: 2023-2024/sourcecode/mod2/lab9/RepositoryBasicsKotlin/app/src/main/java/com/example/repositorybasicskotlin/ui/UserFragment.kt
package com.example.repositorybasicskotlin.ui

import android.os.Bundle
import androidx.fragment.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.repositorybasicskotlin.R
import com.example.repositorybasicskotlin.databinding.FragmentUserBinding
import com.example.repositorybasicskotlin.viewmodel.UserViewModel
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch

class UserFragment : Fragment() {

    private lateinit var binding: FragmentUserBinding

    private val viewModel: UserViewModel by viewModels()
    private val userAdapter by lazy { UserAdapter(UserComparator())}

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentUserBinding.inflate(inflater)

        viewLifecycleOwner.lifecycleScope.launch {
            viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED){
                viewModel.usersList.collectLatest{ users ->
                    userAdapter.submitList(users)
                }
            }
        }

        binding.rvList.apply{
            adapter = userAdapter
            layoutManager = LinearLayoutManager(requireContext())
        }

        return binding.root
    }
}